modules/ank/arrow-ank/src/main/kotlin/arrow/ank/interpreter.kt

```diff
@@ -73,128 +63,141 @@ const val AnkPlaygroundExtension = ":ank:playground:extension"
 
 val ankMacroRegex: Regex = "ank_macro_hierarchy\\((.*?)\\)".toRegex()
 
+sealed class SnippetParserState {
+  data class CollectingCode(val snippet: Snippet) : SnippetParserState()
+  object Searching : SnippetParserState()
+}
+
 fun <F> monadDeferInterpreter(MF: MonadDefer<F>): AnkOps<F> = object : AnkOps<F> {
   override fun MF(): MonadDefer<F> = MF
 
-  override fun printConsole(msg: String): Kind<F, Unit> = MF.delay { println(msg) }
+  override fun printConsole(msg: String): Unit = println(msg)
 
-  override fun Path.containsAnkSnippets(): Kind<F, Boolean> = MF.delay {
-    toFile().useLines { line ->
-      line.find { it.contains(AnkBlock) } != null
+  private fun Path.containsAnkSnippets(): Boolean =
+    toFile().bufferedReader().use {
+      it.lines().anyMatch { s -> s.contains(AnkBlock) || s.contains("ank_macro") }
     }
-  }
 
-  override fun createTargetDirectory(source: Path, target: Path): Kind<F, Path> = MF.delay {
-    source.toFile().copyRecursively(target.toFile(), overwrite = true); target
-  }
-
-  override fun getCandidatePaths(root: Path): Kind<F, Option<Nel<Path>>> = MF.delay {
-    Nel.fromList<Path>(
-      Files.walk(root).filter { Files.isDirectory(it).not() }.filter { path ->
+  override fun <A> Path.withAnkFiles(f: (AnkProcessingContext) -> A): Sequence<A> =
+    Files.walk(this)
+      .filter { Files.isDirectory(it).not() }
+      .filter { path ->
         SupportedMarkdownExtensions.fold(false) { c, ext ->
           c || path.toString().endsWith(ext)
-        }
-      }.collect(Collectors.toList())
-    )
-  }
+        } && path.containsAnkSnippets()
+      }.iterator().asSequence().mapIndexed { index, path ->
+        f(AnkProcessingContext(index, path))
+      }
 
-  override fun readFile(path: Path): Kind<F, String> = MF.delay {
-    String(Files.readAllBytes(path))
+  override fun createTargetDirectory(source: Path, target: Path): Path {
```

Comment:
This is called inside `binding`. Worth making it `Kind<F, Path>`, fix the side-effect, and call it using `bind` on it?

Reply:
I removed for now all polymorphism because the entire path context in which a file is transformed happens inside a `Sequence<A>`. Also I thought originally it was the cause of memory leaks but then found out that the Markdown library was the limitation. We were forced to bring entire files in memory in order to parse the markdown which I also removed. I plan on adding the Kinds back with a different design for the interpreter where you have the `MonadDefer` syntax directly, but I will do that in a separate PR as this one is just to have reliable CI